pulsar-broker/src/main/java/org/apache/pulsar/broker/delayed/bucket/BookkeeperBucketSnapshotStorage.java

```diff
@@ -172,12 +157,25 @@ private CompletableFuture<LedgerHandle> createLedger(String bucketKey, String to
                     if (rc != BKException.Code.OK) {
                         future.completeExceptionally(bkException("Create ledger", rc, -1));
                     } else {
+                        ledgerHandleCache.put(handle.getId(), handle);
                         future.complete(handle);
                     }
                 }, null, metadata);
         return future;
     }
 
+    private CompletableFuture<LedgerHandle> getLedgerHandle(Long ledgerId) {
+        LedgerHandle ledgerHandle = ledgerHandleCache.get(ledgerId);
```

Comment:
There is a race condition, you should consider use `compute` here.

case: read - if -modify